tweaks to gulp tasks

almarklein opened the pull request:
issues: relates to #2140 

This adds functionality to install the js and css to the server directory.

I ended up more than I planned at first, splitting up the "scripts" and "styles" in subtasks which are then run in sequence in "scripts" / "styles". I only realized afterwards that tasks are run in parallel (woops). I solved this with run-sequence. I like how the tasks are split in a couple of simple tasks which can be reused elsewhere. Let me know if this looks good, so I can either move forward with this, or revert to the current approach.

Note that e.g. build.coffee needs changes, but I'll do that when I have a "green light" :)

almarklein commented:
cc @tswicegood @bryevdv

almarklein commented:
It's probably difficult to see from the div, but we basically have:

* scripts:build -> build bokeh.js
* scipts:minifyy -> take bokeh.js and produce bokeh.min.js
* scripts:install -> take everything in `build/` and copy to `server/static/js`
* scripts -> do the three above tasks (in sequence)

same for style.

tswicegood commented on bokehjs/gulp/tasks/develop.coffee:
```diff
@@ -8,8 +8,7 @@ subTasks = ["develop:install-js"]
 gulp.task "develop:watch", ["develop"], ->
-  gulp.watch "#{paths.buildDir.js}/*.js", ["develop:install-js"]
-  gulp.watch "#{paths.buildDir.css}/*css", ["develop:install-js"]
+  gulp.watch "#{paths.coffee.watchSources}", ["develop:install-js"]
```
These should only trigger after successful builds.  With this change, this triggers whenever the `.coffee` files change which would trigger another install before they've been built.

tswicegood replied:
This also doesn't handle CSS.

almarklein replied:
Ah yeah, I was not sure what to do with the develop task yet. How is it used?

tswicegood replied:
Well, my original idea might be outdated. :-)

I originally thought I needed to do the `python setup.py develop --install_js` to get it in a working state.  In that case, the idea was that you'd cd into `bokehjs` and run `gulp develop`.  At that point, everything would built, then installed.  I think you've fixed the root issue with the various new code you've added so we might be able to nix it (yay deleting code!).

tswicegood commented on bokehjs/package.json:
```diff
@@ -44,7 +44,8 @@
     "vinyl-source-stream": "^1.0.0",
-    "vinyl-transform": "^1.0.0"
+    "vinyl-transform": "^1.0.0",
+    "run-sequence": "^1.0.0"
```
This is cool!

tswicegood commented:
Well crap -- I had a message written up, then commented on the diff, then switched back and GH decided I didn't need it any more. :-(

---

This looks ok.  The switch up from using task ordering to `run-sequence` definitely helps with the ordering and makes it possible to remove the `*:minify-only` tasks.  These changes should modify the `default` task so it runs everything appropriately.  I could see that going two ways:

* `gulp` builds everything and watches
* `gulp` builds everyone only and add `gulp watch` that does the build

Given the breakdown of `scripts` and `scripts:watch` and so on, I think I'd go with the default task simply building / installing everything once and `gulp watch` doing that plus watch.

---

Side note on `develop` and `develop:install-js`.  That was written from my perspective of development always happening inside a conda environment, in which case you always have access to your `site-packages`.

That said, if it's simply a matter of copying files around, I'm +1 for doing that in gulp.  The question then becomes should that functionality be removed from `setup.py` with it simply calling out to gulp?  I don't think we should have two ways of getting the JS / CSS installed.

tswicegood commented on bokehjs/gulp/tasks/scripts.coffee:
```diff
@@ -18,22 +19,25 @@ gulp.task "scripts", ->
-minifyTask = ->
-  gulp.src paths.coffee.destination.fullWithPath
+gulp.task "scripts:minify", ->
+  return gulp.src paths.coffee.destination.fullWithPath
```
Style note -- this has mixed explicit / implicit returns.  My preference for the file is implicit, but it should be one style and used throughout.

almarklein replied:
oh sorry, that's an artifact; I added the return in an earlier attempt to make things sequential. will remove.

tswicegood commented on bokehjs/src/coffee/renderer/glyph/line.coffee:
```diff
@@ -9,7 +9,7 @@ class LineView extends Glyph.View
     drawing = false
     @visuals.line.set_value(ctx)
-
+    
```
I don't believe this is related? :-)

almarklein replied:
No, this is me adding whitespace to a file to trigger a rebuild :P

tswicegood replied:
It took whitespace?  What OS are you on?  Normally I can simply save a file to trigger a build.

bryevdv commented:
The python `setup.py` should be responsible for copying things into the python source tree.

almarklein commented:
Ok, changed so that `setup.py --install_js` installs the JS (but does not try any kind of installation). I fixed the build task. Now there's:

* default - show brief message. Or should this actually *do* something? If so, what?
* help - show task list
* build - build + minify
* install - use setup.py to copy to server/static
* watch - watch for changes and do scripts:build followed by install
* scripts:build / scripts:minify - used by other commands
* styles:build / styles:minify - used by other commands

tswicegood commented:
Re: default task -- the pattern I've adopted in other projects is that `gulp` does the standard development mode (generally build, test, watch).  My motivation was to reduce the amount typed the majority of the time.

I'm torn on `scripts` (and `styles`) calling `install` at the end.  If they don't do that, your `build` can be simplified to `runSequence(["scripts", "styles"], "install")` which means scripts and styles can run in parallel.

almarklein commented:
> It took whitespace? What OS are you on? Normally I can simply save a file to trigger a build.

Oh, then its probably the editor that I was using that does not actually save if the file has not changed.

I am new to this gulp stuff, so I have no idea what the convention is :)  Anyway, I made the default do: scripts:build, styles:build, install, watch.

tswicegood commented:
> I am new to this gulp stuff

Wouldn't have guessed that -- you've jumped right in on it! :-)

tswicegood commented on bokehjs/gulp/tasks/default.coffee:
```diff
@@ -1,10 +1,8 @@
-gulp.task "default", defaultTasks, ->
+gulp.task "default", ->
+  console.log('Building BokehJS for developer mode ...')
+  runSequence("scripts:build", "styles:build", "install", "watch")
```
I think this needs to be the `scripts` and `styles` tasks so they get minified.  Actually, if you just call `build` I think you're set.

```coffeescript
runSequence "build", "install", "watch"
```

tswicegood commented:
@almarklein I just tried to run the default command -- there is no `watch` task.  Maybe it's not committed on your side?

almarklein commented:
This is ready for final review. I added the `watch.coffee` that I forgot. I also updated the docs.

bryevdv commented on sphinx/source/docs/dev_guide/building.rst:
```diff
@@ -75,42 +83,36 @@ the complete Bokeh library (see :ref:`devguide_python_setup`).
-standalone libraries. At this point BokehJS can be be as an `AMD module`_.
+To direct Gulp to automatically watch the source tree for changes and
+trigger a recompile if any source file changes. This is especially
+useful together with the ``--splitjs`` mode of the Bokeh server to
```
This is no longer true is it? Does `--splitjs` still work? I thought the browserify change broke it.

almarklein replied:
I don think it works. Is it the intention to have that functionality again?

bryevdv replied:
I'mont sure it is possible, given how browserify works. @tswicegood any thoughts?

tswicegood replied:
Yeah -- that's correct.  There is no splitjs mode right now.

bryevdv commented on sphinx/source/docs/dev_guide/building.rst:
```diff
@@ -211,9 +215,11 @@ not work correctly. The .exe will not be available within the Conda
+.. code-block:: sh
 
-    python bokeh-server --script path\to\<yourapp>.py
+    $ python bokeh-server --script path\to\<yourapp>.py
```
I believe the `$` actually contravene's our recently developed company style guide for docs. Basically states that commands should be presented without adornments like that.

bryevdv replied:
Admittedly, the `$` is all over our docs, we would need to go through and get rid of it I might do that during my upcoming docs work.

bryevdv replied:
ping @kerrywatson1

kerrywatson1 replied:
@bryevdv you are correct, we want to remove each $ and the space after each $ whenever you touch documentation.

tswicegood replied:
FWIW, I thought those were supposed to be in there to show user vs root user (`$` vs `#`).  I'm fine either way, but have been adding `$` throughout when writing docs.

tswicegood replied:
Noted @kerrywatson1 -- I'll start changing them as I come across them.

Do we know what (if any) effect this has on console syntax highlighting in Markdown and ReST?

bryevdv replied:
FWIW (personal opinion, not defined policy): I don't think there is anything in Bokeh that requires running sudo or root any longer so it's not too germane for this in particular. I also think that `$` vs `#` is a very subtle visual distinction and should not be relied upon anyway (better to state explicitly and emphasize: "run this as root") Also I think the benefit of not having it, is to make copy and pasting the commands much easier for users. That said, not putting it has been an adjustment for me too. :) I do think we have to mark `.. code-block: sh` explicitly now but that is fine with me, at least.

almarklein replied:
Ok, I'll get rid of the dollar signs on this page.

tswicegood commented on sphinx/source/docs/dev_guide/building.rst:
```diff
@@ -75,42 +83,34 @@ the complete Bokeh library (see :ref:`devguide_python_setup`).
 .. code-block:: sh
 
-    $ grunt deploy
+    npm install -g gulp
```
Actually, this should be modified to tell it the user to run:

```console
conda install -c javascript gulp
```

That will get them the correct version that's maintained in-house.  You could add a "or install gulp via npm" as an option for people who don't have conda, but that's the best option.

tswicegood commented on sphinx/source/docs/dev_guide/building.rst:
```diff
@@ -119,12 +119,12 @@ Building BokehJS with SBT
         ``setup.py``, so it should be used with caution.
 
-As an alternative to Grunt, you can use `SBT`_ to build BokehJS. To start,
+As an alternative to Gulp, you can use `SBT`_ to build BokehJS. To start,
```
FYI, I don't believe this still works.  @mattpap, can you weigh in?

almarklein replied:
Is there a reason to support it? And even if it still works, is there a reason to document this experimental feature? It add to the clutter IMO.

bryevdv replied:
@mattpap uses it to unify his development process with that for bokeh-scala. If he wants to support it for his own use, that's fine, but we should not mention it in the docs. This entire section ought to be removed.

tswicegood commented on sphinx/source/docs/dev_guide/testing.rst:
```diff
@@ -12,11 +12,11 @@ be viewed at: https://travis-ci.org/bokeh/bokeh.
 To run the just the python unit tests, run the command::
 
-    $ python -c "import bokeh; bokeh.test()"
+    python -c "import bokeh; bokeh.test()"
```
While we're editing this, might as well add the `.. code-block:: python` above (and `sh` for the other lines in this file).

tswicegood replied:
Err, just noticed that this is a shell script executing python, it should be `.. code-block:: python`

almarklein commented:
addressed points.

tswicegood commented:
All looks good to me.  I've run previous versions of this, but haven't run this current code, though.

bryevdv commented on setup.py:
```diff
@@ -364,9 +364,9 @@ def parse_jsargs():
-            sys.exit(1)
+        #if not installing:
+        #    print("Error: Option '--install_js' only valid with 'install', 'develop' or 'sdist', exiting.")
+        #    sys.exit(1)
```
What's going on here?

almarklein replied:
The code I commented prevented the use of only doing `setup.py --install_js`; you could not install js without also installing bokeh-the-package. Having the option to only install js is beneficial, IMO, because its faster and it will always work (i.e. on system Python you wont have permission unless its run under sudo).

I should probbaly *remove* the code rather than comment it though..

bryevdv replied:
OK that seems fine. Yes, please remove it and then merge the PR.
